core/src/main/java/org/elasticsearch/monitor/jvm/JvmInfo.java

```diff
@@ -166,40 +168,104 @@ public static JvmInfo jvmInfo() {
-
-    Map<String, String> systemProperties;
-
-    String[] gcCollectors = Strings.EMPTY_ARRAY;
-    String[] memoryPools = Strings.EMPTY_ARRAY;
-
-    private String onError;
-
-    private String onOutOfMemoryError;
-
-    private String useCompressedOops = "unknown";
+    private final long pid;
+    private final String version;
+    private final String vmName;
+    private final String vmVersion;
+    private final String vmVendor;
+    private final long startTime;
+    private final long configuredInitialHeapSize;
+    private final long configuredMaxHeapSize;
+    private final Mem mem;
+    private final String[] inputArguments;
+    private final String bootClassPath;
+    private final String classPath;
+    private final Map<String, String> systemProperties;
+    private final String[] gcCollectors;
+    private final String[] memoryPools;
+    private final String onError;
+    private final String onOutOfMemoryError;
+    private final String useCompressedOops;
+    private final String useG1GC;
+
+    private JvmInfo(long pid, String version, String vmName, String vmVersion, String vmVendor, long startTime,
+                   long configuredInitialHeapSize, long configuredMaxHeapSize, Mem mem, String[] inputArguments, String bootClassPath,
+                   String classPath, Map<String, String> systemProperties, String[] gcCollectors, String[] memoryPools, String onError,
+                   String onOutOfMemoryError, String useCompressedOops, String useG1GC) {
+        this.pid = pid;
+        this.version = version;
+        this.vmName = vmName;
+        this.vmVersion = vmVersion;
+        this.vmVendor = vmVendor;
+        this.startTime = startTime;
+        this.configuredInitialHeapSize = configuredInitialHeapSize;
+        this.configuredMaxHeapSize = configuredMaxHeapSize;
+        this.mem = mem;
+        this.inputArguments = inputArguments;
+        this.bootClassPath = bootClassPath;
+        this.classPath = classPath;
+        this.systemProperties = systemProperties;
+        this.gcCollectors = gcCollectors;
+        this.memoryPools = memoryPools;
+        this.onError = onError;
+        this.onOutOfMemoryError = onOutOfMemoryError;
+        this.useCompressedOops = useCompressedOops;
+        this.useG1GC = useG1GC;
+    }
 
-    private String useG1GC = "unknown";
+    public JvmInfo(StreamInput in) throws IOException {
+        pid = in.readLong();
+        version = in.readString();
+        vmName = in.readString();
+        vmVersion = in.readString();
+        vmVendor = in.readString();
+        startTime = in.readLong();
+        inputArguments = new String[in.readInt()];
+        for (int i = 0; i < inputArguments.length; i++) {
+            inputArguments[i] = in.readString();
+        }
+        bootClassPath = in.readString();
+        classPath = in.readString();
+        systemProperties = new HashMap<>();
+        int size = in.readInt();
+        for (int i = 0; i < size; i++) {
```

Comment:
This could be `systemProperties = in.readMap(StreamInput::readString, StreamInput::readString);` which I think is a little nicer because you don't have to deal with the size and stuff.